### PR TITLE
Sandwich Attack

### PR DESCRIPTION
**Pull request by Grandthrax**
A general vault sandwich attack. test added to make sure an attacker does make a profit leaching vault profits trading in and out around harvest time.

### CODE REVIEW

File: contracts/test/TestStrategyHighYieldStrat.sol [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+pragma experimental ABIEncoderV2;
+
+import {BaseStrategy, StrategyParams} from "../BaseStrategy.sol";
+
+/*
+ * This Strategy serves as both a mock Strategy for testing, and an example
+ * for integrators on how to use BaseStrategy
+ */
+
+contract TestHighYieldStrategy is BaseStrategy {
```
Comment (fubuloubu) [inline contracts/test/TestStrategyHighYieldStrat.sol:12]: What does this Strategy do that cannot be imitated via `TestStrategy`?

File: tests/functional/strategy/test_attack_vector.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+from brownie import Wei, reverts
```
Comment (fubuloubu) [inline tests/functional/strategy/test_attack_vector.py:1]: Rename this file to ` tests/functional/strategy/test_attacks.py`

File: tests/functional/strategy/test_attack_vector.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import brownie
+
+def test_attack_vector(TestHighYieldStrategy, web3,token, gov, vault,strategist, rando):
```
Comment (fubuloubu) [inline tests/functional/strategy/test_attack_vector.py:4]: rename to `test_sandwich_attack`

File: tests/functional/strategy/test_attack_vector.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    # attacker sees harvest enter tx pool
+    attack_amount = balance
+   
+    #attacker deposits
+    token.approve(vault,attack_amount, {"from": attacker})
+    vault.deposit(attack_amount, {"from": attacker})
+    print('attacker percent of harvest = ', token.balanceOf(strategy)/1e18) 
+    
+    strategy_expected_return = strategy.expectedReturn()
```
Comment (fubuloubu) [inline tests/functional/strategy/test_attack_vector.py:38]: no longer a part of the `BaseStrategy` template, use `Vault.expectedReturn(strategy)` instead

File: tests/functional/strategy/test_attack_vector.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    #attacker withdraws. Pays back loan. and keeps or sells profit
+    vault.withdraw(vault.balanceOf(attacker), {"from": attacker})
+
+    profit = token.balanceOf(attacker) - attack_amount
+    profit_percent = profit/strategy_expected_return
+    print('attacker profit from attack = ', profit/1e18)
+    print('attacker profit share = ', (token.balanceOf(attacker))/attack_amount)
+    print('attacker percent of harvest = ', "{:.2%}".format(profit_percent)) 
+    print("honest_lp share = ", vault.pricePerShare() /1e18)
```
Comment (fubuloubu) [inline tests/functional/strategy/test_attack_vector.py:52]: Remove print statements, use assertions instead

File: tests/functional/strategy/test_attack_vector.py (hunk 2 above)
Comment (Grandthrax) [reply tests/functional/strategy/test_attack_vector.py]: fixed

File: tests/functional/strategy/test_attack_vector.py (hunk 4 above)
Comment (Grandthrax) [reply tests/functional/strategy/test_attack_vector.py]: fixed

File: tests/functional/strategy/test_attack_vector.py (hunk 5 above)
Comment (Grandthrax) [reply tests/functional/strategy/test_attack_vector.py]: fixed

File: tests/functional/strategy/test_attacks.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+    # attacker withdraws. Pays back loan. and keeps or sells profit
+    vault.withdraw(vault.balanceOf(attacker), {"from": attacker})
+
+    profit = token.balanceOf(attacker) - attack_amount
+    profit_percent = profit / strategy_expected_return
+    # print("attacker profit from attack = ", profit / 1e18)
+    # print("attacker profit share = ", (token.balanceOf(attacker)) / attack_amount)
+    # print("attacker percent of harvest = ", "{:.2%}".format(profit_percent))
+    # print("honest_lp share = ", vault.pricePerShare() / 1e18)
```
Comment (fubuloubu) [inline tests/functional/strategy/test_attacks.py:53]:
```suggestion
```

File: tests/functional/strategy/test_attacks.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+def test_sandwich_attack(TestStrategy, web3, token, gov, vault, strategist, rando):
+
+    honest_lp = gov
+    attacker = rando
+    balance = token.balanceOf(honest_lp) / 2
+
+    # seed attacker their funds. Same amount as deposited in vault
+    token.transfer(attacker, balance, {"from": honest_lp})
+    bv = print(token.balanceOf(vault) / 1e18)
```
Comment (fubuloubu) [inline tests/functional/strategy/test_attacks.py:13]:
```suggestion
```

File: tests/functional/strategy/test_attacks.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    vault.withdraw(vault.balanceOf(honest_lp) / 2, {"from": honest_lp})
+    bv = print(token.balanceOf(vault) / 1e18)
```
Comment (fubuloubu) [inline tests/functional/strategy/test_attacks.py:15]:
```suggestion
```

File: tests/functional/strategy/test_attacks.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+def test_sandwich_attack(TestStrategy, web3, token, gov, vault, strategist, rando):
+
+    honest_lp = gov
+    attacker = rando
+    balance = token.balanceOf(honest_lp) / 2
+
+    # seed attacker their funds. Same amount as deposited in vault
+    token.transfer(attacker, balance, {"from": honest_lp})
+    bv = print(token.balanceOf(vault) / 1e18)
+    vault.withdraw(vault.balanceOf(honest_lp) / 2, {"from": honest_lp})
```
Comment (fubuloubu) [inline tests/functional/strategy/test_attacks.py:14]: Why is this withdraw happening?

File: tests/functional/strategy/test_attacks.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    bv = print(token.balanceOf(vault) / 1e18)
+
+    # our actors:
+    # attacker is attacker
+    # strategist is vault owner
+    # honest_lp is a normal lp
+
+    # ------ setup vault ------
```
Comment (fubuloubu) [inline tests/functional/strategy/test_attacks.py:22]: Unnecessary comments
```suggestion
```

File: tests/functional/strategy/test_attacks.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    # we don't use the one in conf because we want no rate limit
+    strategy = strategist.deploy(TestStrategy, vault)
+    vault.addStrategy(
+        strategy, token.totalSupply(), token.totalSupply(), 50, {"from": gov}
+    )
```
Comment (fubuloubu) [inline tests/functional/strategy/test_attacks.py:27]: Just change the rate limit to `2 ** 256 -1` to disable it instead

File: tests/functional/strategy/test_attacks.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    vault.deposit(attack_amount, {"from": attacker})
+
+    # harvest happens
+    strategy.harvest({"from": strategist})
+
+    # attacker withdraws. Pays back loan. and keeps or sells profit
+    vault.withdraw(vault.balanceOf(attacker), {"from": attacker})
+
+    profit = token.balanceOf(attacker) - attack_amount
+    profit_percent = profit / strategy_expected_return
```
Comment (fubuloubu) [inline tests/functional/strategy/test_attacks.py:49]: Do it in terms of the attack capital amount
```suggestion
    profit_percent = profit / attack_amount
```

File: tests/functional/strategy/test_attacks.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+    # ------ setup vault ------
+    # we don't use the one in conf because we want no rate limit
+    strategy = strategist.deploy(TestStrategy, vault)
+    vault.addStrategy(
+        strategy, token.totalSupply(), token.totalSupply(), 50, {"from": gov}
+    )
+
+    strategy.harvest({"from": strategist})
+    strategy_expected_return = token.balanceOf(honest_lp) / 1000
```
Comment (fubuloubu) [inline tests/functional/strategy/test_attacks.py:30]: Don't need this given below

File: tests/functional/strategy/test_attacks.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    vault.withdraw(vault.balanceOf(attacker), {"from": attacker})
+
+    profit = token.balanceOf(attacker) - attack_amount
+    profit_percent = profit / strategy_expected_return
+    # print("attacker profit from attack = ", profit / 1e18)
+    # print("attacker profit share = ", (token.balanceOf(attacker)) / attack_amount)
+    # print("attacker percent of harvest = ", "{:.2%}".format(profit_percent))
+    # print("honest_lp share = ", vault.pricePerShare() / 1e18)
+    # profit percent should be less than 5%
+    assert profit_percent < 0.05
```
Comment (fubuloubu) [inline tests/functional/strategy/test_attacks.py:55]:
```suggestion
    # profit percent should be less than 1%
    assert profit_percent < 0.01
```

File: tests/functional/strategy/test_attacks.py (hunk 9 above)
Comment (Grandthrax) [reply tests/functional/strategy/test_attacks.py:14]: This is so the attacker can attack with same balance as in the vault. and honest_lp needs some money to give to strategy for it to return as profit

Comment (fubuloubu) [reply tests/functional/strategy/test_attacks.py:13]: Weird. Why not account for this externally e.g. in the prior transfer to the attacker?

File: tests/functional/strategy/test_attacks.py (hunk 14 above)
Comment (Grandthrax) [reply tests/functional/strategy/test_attacks.py]: changed to 0.0001

File: tests/functional/strategy/test_attacks.py (hunk 9 above)
Comment (Grandthrax) [reply tests/functional/strategy/test_attacks.py:13]: When vault is deployed in conftest it is seeded with half of the supply of token. Attacker can't equal amount in the vault and also has profit added. Could do it another way if you'd rather. But I thought attacker doubling vault assets was neat. Now we've changed what we're looking for at the end I don't think it matters. As attacker will get same return % on his funds regardless of what he attacks with.

File: tests/functional/strategy/test_attacks.py (hunk 12 above)
Comment (Grandthrax) [reply tests/functional/strategy/test_attacks.py]: done

File: tests/functional/strategy/test_attacks.py (hunk 8 above)
Comment (Grandthrax) [reply tests/functional/strategy/test_attacks.py]: removed

File: tests/functional/strategy/test_attacks.py (hunk 9 above)
Comment (Grandthrax) [reply tests/functional/strategy/test_attacks.py]: Ok redone the maths and changed the way we do it. see latest commit

File: tests/functional/strategy/test_attacks.py (hunk 11 above)
Comment (Grandthrax) [reply tests/functional/strategy/test_attacks.py]: fixed

File: tests/functional/strategy/test_attacks.py (hunk 9 above)
Comment (fubuloubu) [reply tests/functional/strategy/test_attacks.py]: Ahhhh, this makes sense. Also, I think there's a minting function in the test token? If not, you can add it.

Comment (Grandthrax) [reply tests/functional/strategy/test_attacks.py]: Added public mint function to test Token in d98af08

**Comment by dougstorm** [comment]
closing in favor of fix #167

**Pull request closed without merging**